front/lib/api/data_sources.ts

```diff
@@ -211,3 +227,247 @@ export async function augmentDataSourceWithConnectorDetails(
+      ? {
+          prefix: null,
+          content: text,
+          sections: [],
+        }
+      : section || null;
+
+  const nonNullTags = tags || [];
+
+  // Add selection of tags as prefix to the section if they are present.
+  let tagsPrefix = "";
+  ["title", "author"].forEach((t) => {
+    nonNullTags.forEach((tag) => {
+      if (tag.startsWith(t + ":") && tag.length > t.length + 1) {
+        tagsPrefix += `$${t} : ${tag.slice(t.length + 1)}\n`;
+      }
+    });
+  });
+  if (tagsPrefix && generatedSection) {
+    generatedSection.prefix = tagsPrefix;
+  }
+
+  if (!generatedSection) {
+    return new Err(
+      new DustError(
+        "text_or_section_required",
+        "Invalid request body, `text` or `section` must be provided."
+      )
+    );
+  }
+
+  const fullText = sectionFullText(generatedSection);
+
+  const coreAPI = new CoreAPI(apiConfig.getCoreAPIConfig(), logger);
+  const plan = auth.getNonNullablePlan();
+  // Enforce plan limits: DataSource documents count.
+  // We only load the number of documents if the limit is not -1 (unlimited).
+  // the `getDataSourceDocuments` query involves a SELECT COUNT(*) in the DB that is not
+  // optimized, so we avoid it for large workspaces if we know we're unlimited anyway
+  if (plan.limits.dataSources.documents.count != -1) {
+    const documents = await coreAPI.getDataSourceDocuments({
+      projectId: dataSource.dustAPIProjectId,
+      dataSourceId: dataSource.dustAPIDataSourceId,
+      limit: 1,
+      offset: 0,
+    });
+    if (documents.isErr()) {
+      return new Err(
+        new DustError(
+          "core_api_error",
+          "There was an error retrieving the data source."
+        )
+      );
+    }
+
+    if (
+      plan.limits.dataSources.documents.count != -1 &&
+      documents.value.total >= plan.limits.dataSources.documents.count
+    ) {
+      return new Err(
+        new DustError(
+          "data_source_quota_error",
+          `Data sources are limited to ${plan.limits.dataSources.documents.count} ` +
+            `documents on your current plan. Contact support@dust.tt if you want to increase this limit.`
+        )
+      );
+    }
+  }
+
+  // Enforce plan limits: DataSource document size.
+  if (
+    plan.limits.dataSources.documents.sizeMb != -1 &&
+    fullText.length > 1024 * 1024 * plan.limits.dataSources.documents.sizeMb
+  ) {
+    return new Err(
+      new DustError(
+        "data_source_quota_error",
+        `Data sources document upload size is limited to ` +
+          `${plan.limits.dataSources.documents.sizeMb}MB on your current plan. ` +
+          `You are attempting to upload ${fullText.length} bytes. ` +
+          `Contact support@dust.tt if you want to increase it.`
+      )
+    );
+  }
+
+  // Data source operations are performed with our credentials.
+  const credentials = dustManagedCredentials();
+
+  // Create document with the Dust internal API.
+  const upsertRes = await coreAPI.upsertDataSourceDocument({
```

Comment:
@philipperolet is it expected that uploading a document on a static data source from the app does not go through. the upsert queue?